### PR TITLE
Initial DecodingAlgorithm API

### PR DESCRIPTION
**Pull request by matt-gardner**
Here's a first pass at breaking down an encoder/decoder model into separable pieces, so we can have a configurable API for it.  It's just a first pass, so it probably has some things that I did poorly.  @DeNeutoy, I'd appreciate it if you could look over the high-level API design, and see what you think.

The idea is to encapsulate the algorithm that trains or runs the decoder, and keep it separate from the model itself.  Like, if you're building a transition-based parser, and want to try using a learning-to-search algorithm, instead of plain maximum likelihood, this lets you just replace that piece while keeping the same model.

As I said, there are still a bunch of rough edges, so don't bother looking too closely at the specifics of the code.  Just look at the API for now.  Here's some English describing the main ideas:

The model takes a `DecodingAlgorithm` subclass and defines a `DecoderState` subclass.  After doing whatever encoding it wants to do, it creates an `initial_state` instance of the `DecoderState`, then passes that off to the `DecodingAlgorithm`, along with a `decode_step` method that defines the computation done at each step, given the current `DecoderState`.  The `DecodingAlgorithm` runs decoding for a specified number of steps (maybe with a beam, maybe with some other algorithm), then computes a loss (and whatever other outputs you might want from the model), and returns it.  The current `decode` method on the existing `DecodingAlgorithm` class is really simple and pretty hacky (because it's just doing simple maximum likelihood), but it at least demonstrates the basic process of how decoding should proceed.

### CODE REVIEW

File: allennlp/nn/decoding/decoding_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        decoder_state = initial_state
+        finished_states = []
+        for timestep in range(num_steps):
+            decoder_input = self._get_decoder_input(targets,
+                                                    decoder_state.outputs_so_far,
+                                                    timestep,
+                                                    training)
+            action_log_probs, valid_actions, hidden_state = decode_step(decoder_state, decoder_input)
+            if valid_actions is None:
+                decoder_state = decoder_state.update(action_log_probs, hidden_state)
```
Comment (matt-gardner) [inline allennlp/nn/decoding/decoding_algorithm.py]: Because all actions are valid here, we're just replacing the decoder state.  For more complex implementations (and when this is cleaned up) we'll actually create one new state for each valid action, and keep them in a beam, or similar.

File: allennlp/nn/decoding/decoder_state.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+from typing import List
+
+import torch
+
+
+class DecoderState:
+    def __init__(self,
```
Comment (pdasigi) [inline allennlp/nn/decoding/decoder_state.py]: Can you add a docstring for this class? I know this is still WIP, but I'd like to understand why you're storing the things you're storing in this class.

File: allennlp/nn/decoding/decoding_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+               targets: torch.Tensor = None,
+               target_mask: torch.Tensor = None) -> Dict[str, torch.Tensor]:
+        decoder_state = initial_state
+        finished_states = []
+        for timestep in range(num_steps):
+            decoder_input = self._get_decoder_input(targets,
+                                                    decoder_state.outputs_so_far,
+                                                    timestep,
+                                                    training)
+            action_log_probs, valid_actions, hidden_state = decode_step(decoder_state, decoder_input)
```
Comment (pdasigi) [inline allennlp/nn/decoding/decoding_algorithm.py]: Why does `decode_step` need to return `valid_actions`? `action_log_probs` should already contain 0 probs for invalid actions.

**Comment by matt-gardner** [comment]
Ok, this is getting more refined.  I'm now moving past the hacky maximum likelihood implementation, and embracing the fact that we need one `DecoderState` per action.  This should still be reasonably efficient with max marginal likelihood, because we know a priori which actions we should take.  Things are still currently broken, but if you can get some idea of where this is going if you look at it now.  Not saying that you _should_ look at it now - it'll be a lot easier to look at after I've had another couple of hours to hack on it and clean it up.  But I need to go for today.

**Comment by matt-gardner** [comment]
@pdasigi, this is now ready to be looked at.

File: allennlp/nn/decoding/maximum_marginal_likelihood.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    from bad actions to good ones.
+    """
+    def decode(self,
+               initial_state: DecoderState,
+               decode_step: DecoderStep,
+               targets: torch.Tensor = None,
+               target_mask: torch.Tensor = None) -> Dict[str, torch.Tensor]:
+        if targets.dim() != 2:
+            raise NotImplementedError("This implementation cannot yet handle batched inputs")
+        allowed_transitions = self._create_allowed_transitions(targets, target_mask)
```
Comment (pdasigi) [inline allennlp/nn/decoding/maximum_marginal_likelihood.py]: Why is constrained decoding an essential part of `MaximumMarginalLikelihood`? I think it makes more sense for a `DecoderState` to define `allowed_transitions`.

File: tests/data/fields/list_field_test.py [hunk 5]
```diff
@@ -45,11 +48,22 @@ def test_all_fields_padded_to_max_length(self):
...
 
+    def test_nested_list_fields_are_padded_correctly(self):
+        nested_field1 = ListField([LabelField(c) for c in ['a', 'b', 'c', 'd', 'e']])
+        nested_field2 = ListField([LabelField(c) for c in ['f', 'g', 'h', 'i', 'j', 'k']])
+        list_field = ListField([nested_field1, nested_field2])
+        list_field.index(self.vocab)
+        padding_lengths = list_field.get_padding_lengths()
+        assert padding_lengths == {'num_fields': 2, 'list_num_fields': 6}
+        array = list_field.as_array(padding_lengths)
+        numpy.testing.assert_almost_equal(array, [[[0], [1], [2], [3], [4], [-1]],
```
Comment (pdasigi) [inline tests/data/fields/list_field_test.py]: I think you meant to use `assert_array_equal` here?

File: tests/models/encoder_decoders/wikitables_semantic_parser_test.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+from allennlp.common import Params
+from allennlp.data import Dataset, DatasetReader, Instance, Token
+from allennlp.data.fields import Field, LabelField, ListField, TextField
+from allennlp.data.tokenizers import CharacterTokenizer
+from allennlp.data.token_indexers import SingleIdTokenIndexer
+from allennlp.data.dataset_readers.seq2seq import START_SYMBOL, END_SYMBOL
+
+
+@DatasetReader.register("toy_wikitables")
+class ToyWikiTablesDatasetReader(DatasetReader):
```
Comment (pdasigi) [inline tests/models/encoder_decoders/wikitables_semantic_parser_test.py]: Do you plan to use the actual `WikitablesDatasetReader` for tests when it's available?

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 7]
```diff
@@ -311,23 +185,146 @@ def decode(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor
...
+    @overrides
+    def take_step(self,
+                  state: WikiTablesDecoderState,
+                  allowed_actions: Set[int] = None) -> Generator[WikiTablesDecoderState, None, None]:
+        # Outline here: first we'll construct the input to the decoder, which is a concatenation of
+        # an embedding of the decoder input (the last action taken) and an attention over the
+        # question.
+        decoder_hidden, decoder_context = state.hidden_state
+        action = state.action_history[-1] if state.action_history else self._start_index
+        action_input = Variable(decoder_hidden.data.new().resize_(1).long().fill_(action))
```
Comment (pdasigi) [inline allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: You can pass size to `new`.

File: allennlp/nn/decoding/maximum_marginal_likelihood.py (hunk 4 above)
Comment (pdasigi) [reply allennlp/nn/decoding/maximum_marginal_likelihood.py]: Sorry, ignore the previous comment. I misunderstood what you're doing here. I now know that this is not related to constrained decoding.

File: allennlp/nn/decoding/maximum_marginal_likelihood.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        contained types.  So your targets could be strings, for instance, if your ``DecoderStep``
+        and ``DecoderState`` objects know how to handle that.
+        """
+        allowed_transitions: Dict[Tuple[int, ...], Set[int]] = defaultdict(set)
+        for i, target_sequence in enumerate(targets):
+            history: Tuple[int, ...] = ()
+            for j, action in enumerate(target_sequence[1:]):
+                if target_mask[i][j + 1].data[0] == 0:  # +1 because we're starting at index 1, not 0
+                    break
+                allowed_transitions[history].add(action.data[0])
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/nn/decoding/maximum_marginal_likelihood.py]: Again, please ignore the comment above.

File: allennlp/data/fields/list_field.py [hunk 9]
```diff
@@ -47,7 +47,7 @@ def get_padding_lengths(self) -> Dict[str, int]:
         field_lengths = [field.get_padding_lengths() for field in self.field_list]
         padding_lengths = {'num_fields': len(self.field_list)}
         for key in field_lengths[0].keys():
-            padding_lengths[key] = max(x[key] if key in x else 0 for x in field_lengths)
+            padding_lengths['list_' + key] = max(x[key] if key in x else 0 for x in field_lengths)
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/fields/list_field.py]: Done.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 10]
```diff
@@ -35,259 +36,131 @@ class WikiTablesSemanticParser(Model):
...
+        but it's applied to the decoder outputs).
+    decoder_beam_search : ``BeamSearch``
+        When we're not training, this is how we will do decoding.
+    max_decoding_steps : ``int``
+        When we're decoding with a beam search, what's the maximum number of steps we should take?
+        This only applies at evaluation time, not during training.
+    action_namespace : str
+        Parser actions are represented as strings in the data.  This is the namespace in the
+        vocabulary that was used to convert them into integers.
+    action_embedding_dim : int
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Done.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 11]
```diff
@@ -35,259 +36,131 @@ class WikiTablesSemanticParser(Model):
...
-        outputs, compute the input at the current timestep.  Note: This method is agnostic to
-        whether the indices are gold indices or the predictions made by the decoder at the last
-        timestep. So, this can be used even if we're doing some kind of scheduled sampling.
+        encoder_outputs = self._encoder(embedded_input, source_mask)
 
-        If we're not using attention, the output of this method is just an embedding of the input
-        indices.  If we are, the output will be a concatentation of the embedding and an attended
-        average of the encoder inputs.
+        final_encoder_output = encoder_outputs[:, -1]  # (batch_size, encoder_output_dim)
+        with torch.cuda.device_of(final_encoder_output):
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Yep, thanks.  I fixed that one when I tried to actually train this...

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 12]
```diff
@@ -35,259 +36,131 @@ class WikiTablesSemanticParser(Model):
...
-            # (batch_size, input_sequence_length)
-            input_weights = self._decoder_attention(decoder_hidden_state, encoder_outputs, encoder_outputs_mask)
-            # (batch_size, encoder_output_dim)
-            attended_input = weighted_sum(encoder_outputs, input_weights)
-            # (batch_size, encoder_output_dim + target_embedding_dim)
-            return torch.cat((attended_input, embedded_input), -1)
+        if target_action_sequences is not None:
+            # Remove batch dimension and trailing dimension (from ListField[LabelField]]).
+            target_action_sequences = target_action_sequences.squeeze(0).squeeze(-1)
+            target_mask = target_action_sequences > 0
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Hmm, you're right.  But, getting the padding index isn't currently easy, and it's currently guaranteed to be 0, so I'm just going to leave a TODO instead.

File: allennlp/nn/decoding/maximum_marginal_likelihood.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+        We use this to prune the set of actions we consider during decoding, because we only need
+        to score the sequences in ``targets``.
+
+        While the example and types above say that this should be a 2D tensor, the implementation
+        logic is actually more flexible - any sequence of sequences is valid here, with any
+        contained types.  So your targets could be strings, for instance, if your ``DecoderStep``
+        and ``DecoderState`` objects know how to handle that.
+        """
+        allowed_transitions: Dict[Tuple[int, ...], Set[int]] = defaultdict(set)
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/nn/decoding/maximum_marginal_likelihood.py]: Yes, but is there a reason to?  I think the current type is sufficient.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 14]
```diff
@@ -35,259 +36,131 @@ class WikiTablesSemanticParser(Model):
...
-           the sequence           w1  w2  w3  <E> <P> <P>
-           with masks             1   1   1   1   0   0
-           against                l1  l2  l3  l4  l5  l6
-           (where the input was)  <S> w1  w2  w3  <E> <P>
-        """
-        relevant_targets = targets[:, 1:].contiguous()  # (batch_size, num_decoding_steps)
-        relevant_mask = target_mask[:, 1:].contiguous()  # (batch_size, num_decoding_steps)
-        loss = sequence_cross_entropy_with_logits(logits, relevant_targets, relevant_mask)
-        return loss
+        initial_score = Variable(embedded_input.data.new().resize_(1).fill_(0))
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Done.

File: tests/data/fields/list_field_test.py (hunk 5 above)
Comment (matt-gardner) [reply tests/data/fields/list_field_test.py]: I actually prefer `assert_almost_equal` these days.  I think it's nicer about doing type conversions for you.

File: tests/models/encoder_decoders/wikitables_semantic_parser_test.py (hunk 6 above)
Comment (matt-gardner) [reply tests/models/encoder_decoders/wikitables_semantic_parser_test.py]: Of course.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 7 above)
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Done.
